graylog2-server/src/main/java/org/graylog2/log4j/MemoryLimitedCompressingFifoRingBuffer.java

```diff
@@ -104,24 +103,36 @@ private void flush() throws IOException {
      * @param limit    limit the returned entries. This is only a rough estimate and will be rounded to the nearest batch size.
      */
     public synchronized void streamContent(OutputStream outputStream, int limit) {
-        // Round limit to the nearest batch size
-        final AtomicInteger count = new AtomicInteger(limit == 0 ? Integer.MAX_VALUE : limit / BATCHSIZE + 1);
+        limit = limit == 0 ? Integer.MAX_VALUE : limit;
+
+        // The uncompressed currentBatch entries can be limited exactly
+        final int currentBatchSize = currentBatch.size();
+        int skipFromCurrent = limit - currentBatchSize;
+        skipFromCurrent = skipFromCurrent > 0 ? 0 : skipFromCurrent * -1;
+
+        limit -= currentBatchSize;
+
+        // The compressed batches are returned in full
+        final long getCompressedBatches = limit > 0 ? limit / BATCHSIZE + 1 : 0;
+        long skipFromCompressed = getCompressedBatches - compressedRingBuffer.size();
+        skipFromCompressed = skipFromCompressed > 0 ? 0 : skipFromCompressed * -1;
 
         compressedRingBuffer.stream().map(input -> {
                     try {
                         return decompress(input);
                     } catch (IOException e) {
                         throw new RuntimeException(e);
                     }
-                }).takeWhile(b -> count.getAndDecrement() > 1)
+                })
+                .skip(skipFromCompressed)
```

Comment:
I think we can move the skip ahead of the decompression.

Reply:
absolutely :+1: